Add Externally Findable section to info command

ashermancinelli opened the pull request:
Include if/how given package is findable via `spack external find` in `spack info` command.

CC @becker33

ashermancinelli committed:
Add Externally Findable section to info command

becker33 commented on lib/spack/spack/cmd/info.py:
```diff
@@ -155,6 +155,32 @@ def print_text_info(pkg):
+    # if hasattr(pkg, 'determine_spec_details'):
+    #     is_findable = True
+    #     find_method += ', ' if len(find_method) > 0 else ''
+    #     find_method += 'custom'
```
I agree with 'custom' for this if we decide to show this granularity of information

becker33 commented on lib/spack/spack/cmd/info.py:
```diff
@@ -155,6 +155,32 @@ def print_text_info(pkg):
 
+    color.cprint('')
+    color.cprint(section_title('Externally Findable: '))
+    find_method = ''
```
Use a list `find_attributes` here, and then we can print it using `llnl.util.lang.comma_and(find_attributes)`

ashermancinelli commented:
Would it be helpful to just give a boolean value for `Externally Findable`, or are the parts of the package that are findable helpful?

ashermancinelli committed:
Remove unused variable

becker33 commented on lib/spack/spack/cmd/info.py:
```diff
@@ -155,6 +155,32 @@ def print_text_info(pkg):
+    is_findable = False
+    if hasattr(pkg, 'executables'):
+        is_findable = True
+        find_method += 'executables'
```
"executables" is really a different type of information than "variants/versions/custom".  I think we should

1. Change the name to `attributes` instead of `method`
2. not include "executables" in this information (we don't detect the executables, we use them to detect various spec attributes)
3. check for an attribute 'executables' and only do the whole thing if that's set

Then if we later implement detection from libraries or pkgconf files, we can add that to the check for whether it's detectable. I think that a user typing `spack info foo` doesn't really care that we detect foo from an executable, what they care about is "can we find foo and what can we find out about it".

becker33 commented on lib/spack/spack/cmd/info.py:
```diff
@@ -155,6 +155,32 @@ def print_text_info(pkg):
         color.cprint(section_title('Maintainers: ') + mnt)
 
+    color.cprint('')
+    color.cprint(section_title('Externally Findable: '))
```
'Externally detectable' might be better

becker33 reviewed (changes requested):
Made some suggestions, let me know what you think

ashermancinelli committed:
Use more descriptive variable names

No longer use "executables" as a detectable component.
Use "custom" as a detectable component.
Use `comma_and` utility method.

ashermancinelli commented:
Thank you for the careful review @becker33! With this `comma_and` method, I think this might look a little ackward:

```console
$ spack info openmpi
<snip>

Externally Detectable:
    version, variants, and custom

<snip>
```
What do you think?

becker33 commented:
@ashermancinelli were you thinking it would be better to just to a comma-separated list? We could just do `', '.join(find_attributes)` instead.

Also, we should probably not write `custom`, since I realized we're picking up that attribute from the parent class, since that's how the other two are implemented. I think it should say `custom` or maybe just "true" if the other two aren't defined.

ashermancinelli committed:
Use comma delimited detection attributes

ashermancinelli commented:
Yes, imo comma separation alone looks a lot better. Also, you're completely right about the `custom` attribute. I think this looks a lot better.

becker33 reviewed (commented):
@ashermancinelli one thought left before approving this: would it be better to write "True" either way, and then put the details in parenthesis?

```
Externally detectable:
    True (version, variants)
```
vs
```
Externally detectable:
    version, variants
```

That would also make it easier for someone who wants to grep/script around externally detectable packages.

ashermancinelli commented:
I think that's a fair point - this suggests that the section is more clearly answering the question *Is this package externally detectable?* and then secondarily *what parts* are externally detectable.

RE scripting/greping, I have been thinking about a `-o <format>` flag or something like that (eg `spack info openmpi -o yaml`, like kubectl) to address more of my needs, but I think this would be another issue. Your suggestion should suffice.

ashermancinelli committed:
Print "True" even when package detects installation via varaints/version

ashermancinelli committed:
Remove unused import

ashermancinelli committed:
Unit test externally detectable part of spack info

ashermancinelli commented:
Added a unit test for this.
